File: Suspension matlab_org/spring/Tire_suspension_spring/Tire_suspension_spring.m
% Tire + Suspension Coupled System
% ODE-based solver
% 注意!! 這個模型有點偷懶，我在初始狀態直接移動輪胎，讓系統直接從不平衡狀態開始出發

clear; clc; close all;
scriptInfo = dbstack('-completenames');
scriptDir = fileparts(scriptInfo(1).file);
addpath(scriptDir);
params = load_data(fullfile(scriptDir,'data.csv'));
paramNames = fieldnames(params);
for paramIdx = 1:numel(paramNames)
    eval([paramNames{paramIdx} ' = params.(paramNames{paramIdx});']);
end

%% Parameters

ms = tire_ms;      % sprung mass (kg)
mu = tire_mu;      % unsprung mass (kg)

ks = tire_ks;      % suspension spring (N/m)
kt = tire_kt;      % tire stiffness (N/m)

%% Initial conditions

xs0 = tire_xs0;    % body displacement
vs0 = tire_vs0;

xu0 = tire_xu0;    % wheel displacement
vu0 = tire_vu0;

X0 = [xs0; vs0; xu0; vu0];

%% Time span

tspan = tire_tspan;

%% ODE solve

[t,X] = ode45(@(t,X) quarter_car_ode(t,X,ms,mu,ks,kt), tspan, X0);

%% Extract states

xs = X(:,1);
vs = X(:,2);
xu = X(:,3);
vu = X(:,4);

%% Energy (for validation)

KE = 0.5*ms.*vs.^2 + 0.5*mu.*vu.^2;
PE = 0.5*ks.*(xs - xu).^2 + 0.5*kt.*(xu).^2;

E = KE + PE;

E0 = E(1);
E_error = (E - E0)/E0 * 100;

%% Plot

figure;

subplot(3,1,1)
plot(t,xs,'LineWidth',2); hold on;
plot(t,xu,'--','LineWidth',1.5);
grid on;
ylabel('Position (m)');
legend('Sprung','Unsprung');

title('Quarter Car (No Damping) - ODE Solution');

subplot(3,1,2)
plot(t,E,'LineWidth',2);
grid on;
ylabel('Total Energy (J)');

subplot(3,1,3)
plot(t,E_error,'LineWidth',2);
grid on;
xlabel('Time (s)');
ylabel('Energy Error (%)');

function dX = quarter_car_ode(~,X,ms,mu,ks,kt)

xs = X(1);
vs = X(2);
xu = X(3);
vu = X(4);

%% Sprung mass acceleration
as = -(ks/ms)*(xs - xu);

%% Unsprung mass acceleration
au = (ks/mu)*(xs - xu) - (kt/mu)*xu;

%% State derivatives

dX = zeros(4,1);

dX(1) = vs;
dX(2) = as;
dX(3) = vu;
dX(4) = au;

end

File: Suspension matlab/spring/Tire_suspension_spring/Tire_suspension_spring.py
from pathlib import Path
import sys

import matplotlib.pyplot as plt
import numpy as np
from scipy.integrate import solve_ivp

ROOT = next(p for p in Path(__file__).resolve().parents if (p / "suspension_utils.py").exists())
sys.path.insert(0, str(ROOT))

from suspension_utils import load_data, show_plots


def quarter_car_ode(_t, x, ms, mu, ks, kt):
    xs, vs, xu, vu = x
    a_s = -(ks / ms) * (xs - xu)
    a_u = (ks / mu) * (xs - xu) - (kt / mu) * xu
    return [vs, a_s, vu, a_u]


def main() -> None:
    p = load_data(Path(__file__).with_name("data.csv"))
    ms = p["tire_ms"]
    mu = p["tire_mu"]
    ks = p["tire_ks"]
    kt = p["tire_kt"]
    x0 = [p["tire_xs0"], p["tire_vs0"], p["tire_xu0"], p["tire_vu0"]]
    tspan = np.asarray(p["tire_tspan"], dtype=float)
    t_eval = np.linspace(tspan[0], tspan[-1], 1000)

    sol = solve_ivp(quarter_car_ode, (tspan[0], tspan[-1]), x0, t_eval=t_eval, args=(ms, mu, ks, kt))
    t = sol.t
    xs, vs, xu, vu = sol.y

    ke = 0.5 * ms * vs**2 + 0.5 * mu * vu**2
    pe = 0.5 * ks * (xs - xu) ** 2 + 0.5 * kt * xu**2
    energy = ke + pe
    energy_error = (energy - energy[0]) / energy[0] * 100

    fig, axes = plt.subplots(3, 1)
    axes[0].plot(t, xs, linewidth=2, label="Sprung")
    axes[0].plot(t, xu, "--", linewidth=1.5, label="Unsprung")
    axes[0].grid(True)
    axes[0].set_ylabel("Position (m)")
    axes[0].set_title("Quarter Car (No Damping) - ODE Solution")
    axes[0].legend()
    axes[1].plot(t, energy, linewidth=2)
    axes[1].grid(True)
    axes[1].set_ylabel("Total Energy (J)")
    axes[2].plot(t, energy_error, linewidth=2)
    axes[2].grid(True)
    axes[2].set_xlabel("Time (s)")
    axes[2].set_ylabel("Energy Error (%)")
    show_plots()


if __name__ == "__main__":
    main()
